gRPC service for kv storage

kirugan opened the pull request:
Issue: #743

rianhughes commented on grpc/handlers.go:
```diff
@@ -0,0 +1,65 @@
+	}
+}
+
+func (h handlers) handleTxCursor(
+	cur *gen.Cursor,
+	tx *tx,
+	server gen.KV_TxServer,
+) error {
```
```suggestion
func (h handlers) handleTxCursor(cur *gen.Cursor, tx *tx, server gen.KV_TxServer) error {
```

rianhughes commented on grpc/handlers.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	switch cur.Op {
+	case gen.Op_SEEK:
+		key := utils.Flatten(cur.BucketName, cur.K)
+		it.Seek(key)
+	case gen.Op_NEXT:
+		it.Next()
```
Just a question: how are `it.Next()` and `it.Seek(key)` being used here? Do you not need to return the results to the caller?

kirugan replied:
For both methods return value is `bool` and I tried to follow existing proto spec (there is no field that corresponds to wrong iterator). Probably in this case it's better to return error, will check

joshklop replied:
Was reading through the code and almost posted the same question :)

Is it better to return the error here @kirugan?

kirugan replied:
Ok, I'll return error here. The problem with provided proto file is that there is no error => cursor correspondence, probably Rust code will add it later.

fracek replied:
IIRC in rust in case of error we get an empty pair response (with only the cursor id set).

omerfirmak commented on grpc/handlers.go:
```diff
@@ -0,0 +1,65 @@
+func (h handlers) Tx(server gen.KV_TxServer) error {
+	dbTx := h.db.NewTransaction(false)
+	tx := newTx(dbTx)
+	defer tx.cleanup()
```
As a rule, we don't defer functions that can error.

See `db.CloseAndWrapOnError` and please use that to explicitly call cleanup on return statements.

kirugan replied:
Rewrote it with `CloseAndWrap...` function.

kirugan commented on grpc/tx.go:
```diff
@@ -0,0 +1,40 @@
+
+func (t *tx) cleanup() error {
+	for _, it := range t.iterators {
+		it.Close()
```
I updated go to 1.20 in `go.mod` and `README.md` (we already use it in Docker). There is handy function for that - `errors.Join(...)` ❤️

omerfirmak commented on grpc/grpc_test.go:
```diff
@@ -0,0 +1,47 @@
+		if err == io.EOF {
+			fmt.Println("disconnected from server")
+		} else {
+			spew.Dump("error", err)
```
I think require.NoError does this already, no?

kirugan replied:
`spew.Dump(...)` is very convinient pretty printer in Go, I was debugging what kind of grpc errors grpc can I get

omerfirmak commented on node/node.go:
```diff
@@ -149,6 +151,10 @@ func makeHTTP(port uint16, rpcHandler *rpc.Handler, log utils.SimpleLogger) *jso
 	}, log)
 }
 
+func makeGRPC(port uint16, db db.DB, log utils.Logger) *grpc.Server {
+	return grpc.NewServer(port, db, log)
+}
+
 // Run starts Juno node by opening the DB, initialising services.
```
we can inline this?

kirugan replied:
sure, I just wanted to mimic http service with its `makeHTTP(...)`. inlined

omerfirmak reviewed (approved):
I dont think we need to delay this any longer, let's include it in a release and allow apibara to start testing it.

joshklop commented:
Another thought: unit tests for the grpc package could be a good-first-issue.

IronGauntlets commented:
> Another thought: unit tests for the grpc package could be a good-first-issue.

In general, we don't want this to become a practice. I think it is important for the author of the PR to include tests in the PR. I don't think it is fair on others to write tests for code that was authored by someone else as very few people look forward to writing tests. 

We have `codecov/patch` to keep track of this and ideally, we should aim for 80% (it's not always possible that's why it is not a required condition for merging). 1.08% coverage doesn't seem enough.

For now, we can merge it but @kirugan should write tests for it soon.

fracek commented on grpc/handlers.go:
```diff
@@ -0,0 +1,65 @@
+			V:        v,
+			CursorId: cur.Cursor,
+		})
+	default:
```
I believe `Op_FIRST` is not handled.

kirugan replied:
I thought it's the same as `Op_SEEK` + `Op_Current`

fracek commented:
First of all, thank you for this! I played around with this and here's my thoughts:

 * It would be nice to have a `Version()` method that returns the current Juno version. Not because it's _that_ useful to know the version, but because it makes it straight forward to test if the connection is going to work (easier than setting up bidi stream).
 * It looks like the `Open` operation is not supported? It's useful to get a `cursor_id` assigned by the server.
 * Since there are no buckets, is it possible to have a `seek_range` operator that positions at the first element greater than or equal the provided key? That way `seek_range(table_prefix || 0)` is equivalent to positioning to the first element of the table.
   - If that's the behavior of `seek`, then it would be useful to have a `seek_exact` that returns an empty kv if the key doesn't match exactly.
 * How would I implement something like `last`? 
 * I also added some smaller comments to the discussion.

kirugan commented:
@fracek 

> It would be nice to have a Version() method ...

Implemented 

> It looks like the Open operation is not supported? It's useful to get a cursor_id assigned by the server.

Initially I thought client should send `cursorId` and if doesn't exist server creates new iterator for new `cursorId`. I've changed this logic, added `open` operation

> If that's the behavior of seek, then it would be useful to have a seek_exact that returns an empty kv if the key doesn't match exactly.

yep, current `seek` works the way you described `seek_range` operation. I've also added `seek_exact`

> How would I implement something like last?

The only way to do it right now is to iterate over all keys with same prefix until you reach next prefix key. I know it's not super convenient, I'll think how to fix it.

> I also added some smaller comments to the discussion.

Thanks

kirugan commented:
I'll merge this PR. The main functionality is implemented. All subsequent edits and features will be added to other pr
